### DISCUSSION TITLE
OpenSenseMap based API endpoint

### DISCUSSION THREAD

**Discussion by CargoBikoMeter** [technologiestiftung/stadtpuls#6] [Ideas]
I would be nice, if the application could work with OpenSenseMap based paylaod decoder for TTN. Then I could send the data of my already existing IoT sensors to the stadpuls api to. TTN console allows to define multiple Webhooks for a device but only one payload decoder function for a device. If you could provide a second API endpoint, that would be very nice. In the stadtpuls sensor configuration menu it should be possible to assign the OpenSenseMap internal hex values to stadtpuls measurement values.

My OpenSenseMap Payload Decoder:
```
function Decoder(b, port) {
 
 if (port === 1) {
  var voltage = (b[0] | b[1]<<8 | (b[1] & 0x80 ? 0xFF<<16 : 0)) / 1000;
  var temperature = (b[2] | b[3]<<8 | (b[3] & 0x80 ? 0xFF<<16 : 0)) / 100;
  //var temp = ((b[2] & 0x80 ? 0xFFFF<<16 : 0) | b[2]<<8 | b[3]) / 100;
  var humidity = (b[4] | b[5]<<8 | (b[5] & 0x80 ? 0xFF<<16 : 0)) / 100;
  var pressure = (b[6] | b[7]<<8 | (b[7] & 0x80 ? 0xFF<<16 : 0));
  var p25 = (b[8] | b[9]<<8 | (b[9] & 0x80 ? 0xFF<<16 : 0)) / 100;
  var p10 = (b[10] | b[11]<<8 | (b[11] & 0x80 ? 0xFF<<16 : 0)) / 100;
  
    return {
    //opensensemap.org 
    '61115f385f5cac001b3bb4f8': temperature,
    '61115f385f5cac001b3bb4f7': humidity,
    '61115f385f5cac001b3bb4f6': pressure,
    '61115f385f5cac001b3bb4f4': p10,
    '61115f385f5cac001b3bb4f3': p25,
    '61115f385f5cac001b3bb4f5': voltage
  };
 }
}
```

Regards,
Adlerkiez-IoT from Berlin Adlershof

**Comment by Esshahn** [comment]
Hey CargoBikoMeter, thank you for the idea and giving Stadtpuls a try. We will discuss your idea in the team!

**Comment by ff6347** [comment]
You could additionally add data like our formatter returns it and API expects it. I think the opensensemap part should be ignored by api.stadtpuls.com. Like so: 

    return {
    //opensensemap.org 
    '61115f385f5cac001b3bb4f8': temperature,
    '61115f385f5cac001b3bb4f7': humidity,
    '61115f385f5cac001b3bb4f6': pressure,
    '61115f385f5cac001b3bb4f4': p10,
    '61115f385f5cac001b3bb4f3': p25,
    '61115f385f5cac001b3bb4f5': voltage
     // Stadtpuls 
    data: {
      measurements: [temperature, humidity, /* and all the others */]
    },
    };

If it throws an error let me know. It's a small change in the schema.

**Reply by ff6347** [reply]
Currently we only display the first value received in `measurements` but you can send more then one value and stadtpuls stores it. You can get them all through the api. 

We're still working on a good visualization and UX for compound sensors

**Reply by ff6347** [reply]
@CargoBikoMeter I just tested to push some data with additional properties to our ttn endpoint. Our API doesn't care if we have `additionalProperties` [see the source](https://github.com/technologiestiftung/stadtpuls-api/blob/c5024a6e3b7cfc434e8992cd8e1f3317e1ef478f/src/integrations/ttn.ts#L87-L89). So you can use your payload formater for both if opensensemap is okay with that too

**Reply by CargoBikoMeter** [reply]
Hi, found today this nice project https://revspace.nl/Sensor-data-bridge, which allows the forwarding of TTN device data to different databases. So I will setup that forwarder on my system, which allows me to send the data to stadtpuls and opensensmap with no modifications of the opensensmap API.
Maybe you could install that forwarder at the stadtpuls.com server too, to allow the forwarding of sensor data from TTN to different databases for the user of stadtpuls.

**Reply by ff6347** [reply]
> Hi, found today this nice project https://revspace.nl/Sensor-data-bridge, which allows the forwarding of TTN device data to different databases.

Nice find. Thanks. We will take a look. 

> Maybe you could install that forwarder at the stadtpuls.com server too, to allow the forwarding of sensor data from TTN to different databases for the user of stadtpuls.

We actually don't want to get in the position to maintain outgoing traffic, only incoming. Even if we start doing this, we would take a more generic approach. Like having server send events people can subscribe to our like MQTT topics.

**Comment by CargoBikoMeter** [comment]
Hi,

did I understand the Stadtpuls API correctly: The API does not support key/value JSON pairs for dedicated measurement values? I would like to send the data from my logstash based system by using the logstash filter plugin:
       mutate {
            add_field => { "[measurements][pressure]" => "%{pressure}" }
       }
or via the output http plugin:
    mapping => ["pressure","%{pressure}"]

By using the output plugin I could avoid all the staff which is coming from the TTN system, only the required data for the API would be send. But the mapping function does not support nested JSON arrays.

Because I use a dedicated Uplink-URL for each sensor (e.g. https://api.stadtpuls.com/api/v3/sensors/72/records), it would be easier to send a key/value pair.

What I understand is, I have to encode the measurement JSON without a dedicated key identifier, only the values, why?

Could you provide a curl based command line for posting measurements to the API in the documentation https://stadtpuls.com/docs?

**Reply by ff6347** [reply]
Hey. Yes we don't store key value pairs. We want to be agnostic of any "name" for the records. The type is defined by the sensor type.

You can send records like this via HTTP:

```bash
curl --location --request POST 'https://api.stadtpuls.com/api/v3/sensors/<YOUR SENSOR ID>/records' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer <YOUR TOKEN>' \
--data-raw '{
    "measurements": [
        1,
        2,
        3
    ]
}'
```

**Reply by ff6347** [reply]
If you want one single endpoint you could use the TTN endpoint. But then your POST body needs to have the `device_id` in it and your sensors integration needs to be of type `TTN`.
```bash

curl --location --request POST 'https://api.stadtpuls.com/api/v3/integrations/ttn/v3' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer <YOUR TOKEN>' \
--data-raw '{
    "simulated": true,
    "end_device_ids": {
        "application_ids": {
            "application_id": "foo"
        },
        "device_id": "<YOUR DEVICE ID LIKE SET IN STADTPULS>COM>"
    },
    "received_at": "2022-02-09T10:40:03.338Z", //<--- SET THIS TO DEFINE THE recorded_at field
    "uplink_message": {
        "decoded_payload": {
            "foo": "bah",
            "measurements": [ // <-- THIS  measurements go
                1,
                2,
                3
            ],
            "bytes": [
                1,
                2,
                3
            ]
        },
        "locations": {
            "user": {
                "latitude": 13,
                "longitude": 52,
                "altitude": 23,
                "source": "SOURCE_REGISTRY"
            }
        }
    }
}'
```

**Comment by Esshahn** [comment]
Hallo @CargoBikoMeterl, wir hatten am Freitag in der offenen Werkstatt zu Pipedream gesprochen. Kannst Du mir eine Mail schreiben an ingo.hinterding@ts.berlin?
